R/programming.R

```diff
@@ -0,0 +1,76 @@
+is.AsIs = function(x) {
+  inherits(x, "AsIs")
+}
+rm.AsIs = function(x) {
+  cl = oldClass(x)
+  oldClass(x) = cl[cl!="AsIs"]
+  x
+}
+list2lang = function(x) {
+  if (!is.list(x))
+    stop("'x' must be a list")
+  if (is.AsIs(x))
+    return(rm.AsIs(x))
+  asis = vapply(x, is.AsIs, FALSE)
+  char = vapply(x, is.character, FALSE)
+  to.name = !asis & char
+  if (any(to.name)) { ## turns "my_name" character scalar into `my_name` symbol, for convenience
+    if (any(non.scalar.char <- vapply(x[to.name], length, 0L)!=1L)) {
+      stop("Character objects provided in the input are not scalar objects, if you need them as character vector rather than a name, then wrap each into 'I' call: ",
+           paste(names(non.scalar.char)[non.scalar.char], collapse=", "))
+    }
+    x[to.name] = lapply(x[to.name], as.name)
+  }
+  if (isTRUE(getOption("datatable.enlist", TRUE))) { ## recursively enlist for nested lists, see note section in substitute2 manual
+    islt = vapply(x, is.list, FALSE)
+    to.enlist = !asis & islt
+    if (any(to.enlist)) {
+      x[to.enlist] = lapply(x[to.enlist], enlist)
+    }
+  }
+  if (any(asis)) {
+    x[asis] = lapply(x[asis], rm.AsIs)
+  }
+  x
+}
+enlist = function(x) {
+  if (!is.list(x))
+    stop("'x' must be a list")
+  if (is.AsIs(x))
+    return(rm.AsIs(x))
+  as.call(c(quote(list), list2lang(x)))
+}
+
+substitute2 = function(expr, env) {
+  if (missing(env)) {
+    stop("'env' must not be missing")
+  } else if (is.null(env)) {
+    # null is fine, will be escaped few lines below
+  } else if (is.environment(env)) {
+    env = as.list(env, all.names=TRUE, sorted=TRUE)
```

Comment:
Interesting question...

TL;DR
works as documented.

----

Following chunk of code presents current difference vs base R substitute.
```r
e = new.env()
f = function() {message("f"); "V1"}
makeActiveBinding("x", f, e)
e$x
#f
#[1] "V1"
substitute2(list(x), e)
#list(function() {message("f"); "V1"})
substitute(list(x), e)
#f
#list("V1")
```
To have best consistency to base R substitute, we should convert list to an environment rather than environment to a list.
Another benefit of that would be to avoid repeating C `chmatch` (one time for each nested, named list) that we now use to compare names from `expr` with names from `env`. When having environment object we can use `findVarInFrame3`. Although now sure which one will be faster.
On the other hand having environment object would be more complex to pre-process (character to symbol, list to list call, recursively), as we would need to take care of the promises to not evaluate them (see test 2.820).
In the end, we somehow document it by: `env` is _List, or an environment that will be coerced to list, from which variables will be taken to inject into `expr`_.
Considering the use case is rather uncommon, I wouldn't much bother now about addressing it. Did you actually use active bindings that you would like to use passing in an enviroment for substitution?

Reply:
Thanks for your detailed reply! Just curious about the consistency, personally, I don't have any use case of active bindings.

I think the documentation looks good to make it clear.